keyboards/unicomp/mini_m/minirazz/matrix.c

```diff
@@ -0,0 +1,159 @@
+/* Copyright 2023 Purdea Andrei
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "quantum.h"
+#include <hardware/spi.h>
+#include <hardware/gpio.h>
+#include <hardware/structs/pads_qspi.h>
+
+matrix_row_t previous_matrix[MATRIX_ROWS];
+
+static inline void setPinInputHigh_atomic(pin_t pin) {
+    ATOMIC_BLOCK_FORCEON {
+        setPinInputHigh(pin);
+    }
+}
+
+static inline void setPinOutput_writeLow(pin_t pin) {
+    ATOMIC_BLOCK_FORCEON {
+        setPinOutput(pin);
+        writePinLow(pin);
+    }
+}
+
+void matrix_init_custom(void) {
+    for (int i=0; i<=7; i++)
+    {
+        setPinInputHigh(i);
+    }
+    for (int i=8; i<=19; i++)
+    {
+        writePinLow(i);
+        setPinInputHigh(i);
+    }
+    spi_init(spi1, 5000000);
+    spi_cpol_t cccpol = 1;
+    spi_cpha_t cccpha = 0;
+    spi_set_format(spi1, 8, cccpol, cccpha, SPI_MSB_FIRST);
+    gpio_set_function(26, GPIO_FUNC_SPI); // SCK
+    gpio_set_function(24, GPIO_FUNC_SPI); // RX (MISO)
+    writePinLow(27);
+    setPinOutput(27);
+
+    for (int gpio=0;gpio<NUM_BANK0_GPIOS;gpio++) {
+        gpio_set_drive_strength(gpio, GPIO_DRIVE_STRENGTH_2MA);
+    }
```

Comment:
Here's a dump of all GPIO settings if I comment out that line, we're talking about the "DRIVE" column here:
```
GPIO0    : OD=0 IE=1 DRIVE=0 PUE=1 PDE=0 SCHMITT=1 SLEWFAST=0
GPIO1    : OD=0 IE=1 DRIVE=0 PUE=1 PDE=0 SCHMITT=1 SLEWFAST=0
GPIO2    : OD=0 IE=1 DRIVE=0 PUE=1 PDE=0 SCHMITT=1 SLEWFAST=0
GPIO3    : OD=0 IE=1 DRIVE=0 PUE=1 PDE=0 SCHMITT=1 SLEWFAST=0
GPIO4    : OD=0 IE=1 DRIVE=0 PUE=1 PDE=0 SCHMITT=1 SLEWFAST=0
GPIO5    : OD=0 IE=1 DRIVE=0 PUE=1 PDE=0 SCHMITT=1 SLEWFAST=0
GPIO6    : OD=0 IE=1 DRIVE=0 PUE=1 PDE=0 SCHMITT=1 SLEWFAST=0
GPIO7    : OD=0 IE=1 DRIVE=0 PUE=1 PDE=0 SCHMITT=1 SLEWFAST=0
GPIO8    : OD=0 IE=1 DRIVE=0 PUE=1 PDE=0 SCHMITT=1 SLEWFAST=0
GPIO9    : OD=0 IE=1 DRIVE=0 PUE=1 PDE=0 SCHMITT=1 SLEWFAST=0
GPIO10   : OD=0 IE=1 DRIVE=0 PUE=1 PDE=0 SCHMITT=1 SLEWFAST=0
GPIO11   : OD=0 IE=1 DRIVE=0 PUE=1 PDE=0 SCHMITT=1 SLEWFAST=0
GPIO12   : OD=0 IE=1 DRIVE=0 PUE=1 PDE=0 SCHMITT=1 SLEWFAST=0
GPIO13   : OD=0 IE=1 DRIVE=0 PUE=1 PDE=0 SCHMITT=1 SLEWFAST=0
GPIO14   : OD=0 IE=1 DRIVE=0 PUE=1 PDE=0 SCHMITT=1 SLEWFAST=0
GPIO15   : OD=0 IE=1 DRIVE=0 PUE=1 PDE=0 SCHMITT=1 SLEWFAST=0
GPIO16   : OD=0 IE=1 DRIVE=0 PUE=1 PDE=0 SCHMITT=1 SLEWFAST=0
GPIO17   : OD=0 IE=1 DRIVE=0 PUE=1 PDE=0 SCHMITT=1 SLEWFAST=0
GPIO18   : OD=0 IE=1 DRIVE=0 PUE=1 PDE=0 SCHMITT=1 SLEWFAST=0
GPIO19   : OD=0 IE=1 DRIVE=0 PUE=1 PDE=0 SCHMITT=1 SLEWFAST=0
GPIO20   : OD=0 IE=1 DRIVE=0 PUE=0 PDE=0 SCHMITT=0 SLEWFAST=0
GPIO21   : OD=0 IE=1 DRIVE=0 PUE=0 PDE=0 SCHMITT=0 SLEWFAST=0
GPIO22   : OD=0 IE=1 DRIVE=0 PUE=0 PDE=0 SCHMITT=0 SLEWFAST=0
GPIO23   : OD=0 IE=1 DRIVE=0 PUE=0 PDE=0 SCHMITT=0 SLEWFAST=0
GPIO24   : OD=0 IE=1 DRIVE=1 PUE=0 PDE=1 SCHMITT=1 SLEWFAST=0
GPIO25   : OD=0 IE=1 DRIVE=0 PUE=0 PDE=0 SCHMITT=0 SLEWFAST=0
GPIO26   : OD=0 IE=1 DRIVE=1 PUE=0 PDE=1 SCHMITT=1 SLEWFAST=0
GPIO27   : OD=0 IE=1 DRIVE=0 PUE=0 PDE=0 SCHMITT=0 SLEWFAST=0
GPIO28   : OD=0 IE=0 DRIVE=1 PUE=0 PDE=1 SCHMITT=1 SLEWFAST=0
GPIO29   : OD=0 IE=0 DRIVE=1 PUE=0 PDE=1 SCHMITT=1 SLEWFAST=0
```

So it is necessary at the very least for GPIO26, with is the SPI clock output. Would you prefer I call `palSetLineMode(GP26, PAL_MODE_OUTPUT_PUSHPULL | PAL_RP_PAD_DRIVE2);` instead of the above for loop?
The for loop makes me feel safer, that any future defaults changing would not affect my code.

Reply:
That's probably because GPIO26 ends up being configured by the pico-sdk code, not ChibiOS.

BTW, for the QMK `spi_master` code the settings for SPI pins on RP2040 are defined in `platforms/chibios/chibios_config.h`:
```c
#    define SPI_SCK_PAL_MODE (PAL_MODE_ALTERNATE_SPI | PAL_RP_PAD_SLEWFAST | PAL_RP_PAD_DRIVE4)
#    define SPI_MOSI_PAL_MODE SPI_SCK_PAL_MODE
#    define SPI_MISO_PAL_MODE SPI_SCK_PAL_MODE
```
(and they are apparently not easily changeable from that, which may become an additional problem for you when switching to the QMK SPI API).

Reply:
Note also that the reset value of the PADS_BANK0 register configures 4mA drive strength. (and this is why some untouched GPIOs also have drive strength of 1=4mA). So the current QMK default is unlike the hardware default, so I wouldn't be surprised it this ended up changing in the future.